refactor(common): add defaults to new generic parameters

crisbeto opened the pull request:
This is a follow-up to #33997 where some new generic parameters were added without defaults which is technically a breaking change. These changes add the defaults.

JoostK commented on packages/common/src/directives/ng_for_of.ts:
```diff
@@ -11,7 +11,7 @@ import {Directive, DoCheck, EmbeddedViewRef, Input, IterableChangeRecord, Iterab
  * @publicApi
  */
-export class NgForOfContext<T, U extends NgIterable<T>> {
+export class NgForOfContext<T, U extends NgIterable<T> = any> {
```
We should be able to use `NgIterable<T>` as default type, here and in all other places:

```suggestion
export class NgForOfContext<T, U extends NgIterable<T> = NgIterable<T>> {
```

crisbeto replied:
Done.

JoostK reviewed (commented):
Could you also update the template type check type definitions?

IgorMinar reviewed (commented):
Could we set the default to be `unknown` rather than `any`?

crisbeto committed:
refactor(common): add defaults to new generic parameters

This is a follow-up to #33997 where some new generic parameters were added without defaults which is technically a breaking change. These changes add the defaults.

crisbeto commented:
Done.

gkalpak commented on packages/common/src/directives/ng_if.ts:
```diff
@@ -149,7 +149,7 @@ import {Directive, EmbeddedViewRef, Input, TemplateRef, ViewContainerRef, ɵstri
  */
 @Directive({selector: '[ngIf]'})
-export class NgIf<T> {
+export class NgIf<T = unknown> {
```
Have you verified that this avoids the breaking change?

crisbeto replied:
I don't have a way of verifying it, but it should, given that the error is about a required parameter not being defined.

gkalpak replied:
I used the [build artifacts from this PR][1] to test on `ngcc-validation` (where we got a [failure due to #33997][2]) and it passed: angular/ngcc-validation#656. But then I realized that it only exercises the `NgForOfContext` change (whose type does not default to `unknown`) 🤷‍♂ 

Since I don't know how to verify this either, let's assume it's fine until proven otherwise :grin:

[1]: https://circleci.com/gh/angular/angular/551837#artifacts/containers/0
[2]: https://circleci.com/gh/angular/ngcc-validation/2188#tests/containers/1

mhevery commented:
[presubmit](http://test/OCL:283605236:BASE:283605318:1575407045550:81a4316)
